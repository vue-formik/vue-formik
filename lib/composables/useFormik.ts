import { computed, reactive, toRaw, watch, ref, type UnwrapRef } from "vue";
import { ObjectSchema as YupObjectSchema } from "yup";
import { ObjectSchema as JoiObjectSchema } from "joi";
import { clearReactiveObject, getNestedValue, updateNestedProperty } from "@/helpers";
import type { FormikHelpers, FormikOnSubmit, FormikValidationSchema, FormikMode } from "@/types";

/**
 * Type for form field events
 */
type FieldElement = HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement;

/**
 * Enhanced Formik-like form management hook for Vue
 */
const useFormik = <T extends object>({
  initialValues,
  validationSchema,
  onSubmit,
  validateOnMount = true,
  preventDefault = true,
  mode = "YUP",
}: {
  initialValues: T;
  validateOnMount?: boolean;
  preventDefault?: boolean;
  onSubmit?: FormikOnSubmit<T>;
  validationSchema?: FormikValidationSchema<T>;
  mode?: FormikMode;
}) => {
  // Refs for tracking form state
  const isSubmitting = ref(false);
  const initialValuesRef = ref<T>({ ...initialValues });
  const isValidating = ref(false);
  const submitCount = ref(0);

  /**
   * Enhanced validation function with better error handling
   */
  const validate = (values: T): Partial<Record<keyof T, unknown>> => {
    const validationErrors: Partial<Record<keyof T, unknown>> = {};

    if (!validationSchema) {
      return validationErrors;
    }

    if (mode === "JOI") {
      const vSchema = validationSchema as JoiObjectSchema;
      const { error: e } = vSchema.validate(values, { abortEarly: false });
      const err = e as {
        details?: Array<{
          message: string;
          context: {
            label: string;
          }
        }>
      };
      if (err?.details?.length) {
        console.log(err.details)
        err.details.forEach(({ context, message }) => {
          console.log(
            `key: ${context.label}, message: ${message}`,
            validationErrors,
          )
          updateNestedProperty(validationErrors as Record<string, unknown>, context.label, message);
        });
      }
    } else if (mode === "YUP") {
      try {
        const vSchema = validationSchema as YupObjectSchema<T>;
        vSchema.validateSync(values, { abortEarly: false });
      } catch (e) {
        const err = e as { inner?: Array<{ path: string; message: string }> };
        if (err?.inner?.length) {
          err.inner.forEach(({ path, message }) => {
            updateNestedProperty(validationErrors as Record<string, unknown>, path, message);
          });
        }
      }
    } else if (typeof validationSchema === "object") {
      Object.entries(validationSchema).forEach(([key, rules]) => {
        if (typeof rules === "function") {
          const value = getNestedValue(values as Record<string, unknown>, key);
          const error = rules(value);
          if (error) {
            updateNestedProperty(validationErrors as Record<string, unknown>, key, error);
          }
        }
      });
    } else {
      console.error("Invalid validation schema provided");
    }

    return validationErrors;
  };

  // Reactive form state
  const values = reactive<T>({ ...initialValues });
  const errors = reactive<Partial<Record<keyof T, unknown>>>({});
  const touched = reactive<Partial<Record<keyof T, unknown>>>({});

  // Computed properties for form state
  const isValid = computed(() => Object.keys(errors).length === 0);
  const isDirty = computed(() => JSON.stringify(values) !== JSON.stringify(initialValuesRef.value));

  /**
   * Enhanced state setters with type safety
   */
  const setValues = (newValues: Partial<T>) => {
    Object.assign(values, newValues);
  };

  const setErrors = (newErrors: Partial<Record<keyof T, unknown>>) => {
    clearReactiveObject(errors);
    Object.assign(errors, newErrors);
  };

  const setTouched = (newTouched: Partial<Record<keyof T, unknown>>) => {
    clearReactiveObject(touched);
    Object.assign(touched, newTouched);
  };

  /**
   * Enhanced reset function with better type safety
   */
  const reset = ({
    values: newValues,
    keepTouched = false,
  }: {
    values?: Partial<T>;
    keepTouched?: boolean;
  } = {}) => {
    const resetValues = newValues || { ...initialValuesRef.value };
    setValues({ ...resetValues });
    Object.assign(initialValuesRef.value, resetValues);
    if (!keepTouched) {
      clearReactiveObject(touched);
    }
    submitCount.value = 0;
  };

  /**
   * Enhanced field manipulation functions
   */
  const setFieldValue = (field: string, value: unknown) => {
    updateNestedProperty(values as Record<string, unknown>, field, value);
  };

  const setFieldTouched = (field: string, isTouched: boolean) => {
    updateNestedProperty(touched as Record<string, unknown>, field, isTouched);
  };

  const setSubmitting = (value: boolean) => {
    isSubmitting.value = value;
  };

  /**
   * Enhanced submit handler with better error handling
   */
  const handleSubmit = (e?: Event) => {
    if (typeof onSubmit !== "function") return;

    if (preventDefault && e) {
      e.preventDefault();
    }

    submitCount.value++;
    setSubmitting(true);
    isValidating.value = true;

    try {
      const validationErrors = validate(toRaw(values) as T);
      setErrors(validationErrors);

      if (Object.keys(validationErrors).length === 0) {
        onSubmit(
          toRaw(values) as T,
          {
            reset,
            setErrors,
            setValues,
            setSubmitting,
            setTouched,
            setFieldValue,
            setFieldTouched,
          } as FormikHelpers<T>,
        );
      }
    } catch (error) {
      console.error("Form submission error:", error);
    } finally {
      isValidating.value = false;
    }
  };

  /**
   * Enhanced field event handlers
   */
  const handleFieldBlur = (e: FocusEvent) => {
    const target = e.target as FieldElement;
    setFieldTouched(target.name, true);
  };

  const handleFieldChange = (e: Event) => {
    const target = e.target as HTMLInputElement;
    const value = target.type === "checkbox" ? target.checked : target.value;

    setFieldValue(target.name, value);
    setFieldTouched(target.name, true);
  };

  /**
   * Enhanced validation handling
   */
  const performValidation = () => {
    isValidating.value = true;
    const validationErrors = validate(toRaw(values) as T);
    clearReactiveObject(errors);
    Object.assign(errors, validationErrors);
    isValidating.value = false;
  };

  if (validateOnMount) {
    performValidation();
  }

  // Debounced validation on value changes
  watch(
    () => values,
    () => {
      performValidation();
    },
    { deep: true },
  );

  /**
   * Enhanced field state getters
   */
  const hasFieldError = (field: string): boolean => {
    const errorValue = getNestedValue(errors as Record<string, unknown>, field);
    const touchedValue = getNestedValue(touched as Record<string, unknown>, field);
    return Boolean(errorValue && touchedValue);
  };

  const getFieldError = (field: string) => {
    if (hasFieldError(field)) {
      return getNestedValue(errors as Record<string, unknown>, field);
    } else {
      return "";
    }
  };

  const getFieldValue = <K extends keyof UnwrapRef<T>>(field: K | string) => {
    return getNestedValue(values as Record<string, unknown>, field as string);
  };

  /**
   * Computed field handlers for component binding
   */
  const fieldHandlers = computed(() => ({
    onBlur: handleFieldBlur,
    onChange: handleFieldChange,
  }));

  return {
    // Form State
    values,
    errors,
    touched,
    isValid,
    isDirty,
    isSubmitting,
    isValidating,
    submitCount,

    // Field Handlers
    fieldHandlers,

    // State Setters
    setValues,
    setErrors,
    setTouched,
    setFieldValue,
    setFieldTouched,
    setSubmitting,

    // Form Actions
    reset,
    handleSubmit,

    // Field Event Handlers
    handleFieldBlur,
    handleFieldChange,

    // Field State Getters
    hasFieldError,
    getFieldError,
    getFieldValue,
  };
};

export default useFormik;
