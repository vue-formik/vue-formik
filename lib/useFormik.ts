import { computed, reactive, toRaw, watch } from 'vue'
import { ObjectSchema } from 'yup'
import type { FormikHelpers, IUseFormik, ValidationRule } from '@/types'
import { getNestedValue, updateNestedProperty } from '@/helpers'

const useFormik = <T extends object>({
  initialValues,
  validationSchema,
  onSubmit,
}: IUseFormik<T>) => {
  const validate = (values: T): Partial<Record<keyof T, string>> => {
    const validationErrors: Partial<Record<keyof T, string>> = {};

    if (validationSchema instanceof ObjectSchema) {
      try {
        const vSchema = validationSchema as ObjectSchema<T>;
        vSchema.validateSync(values, { abortEarly: false });
      } catch (e) {
        const err = e as never as { inner: { path: string; message: string }[] };
        if (typeof err === "object" && Array.isArray(err?.inner)) {
          err.inner.forEach((error) => {
            validationErrors[error.path as keyof T] = error.message;
          });
        }
      }
    } else {
      for (const key in validationSchema) {
        const value = values[key as keyof T];
        const rule = validationSchema[key as keyof T] as ValidationRule<T[keyof T]>;

        if (typeof rule === 'function') {
          const error = rule(value);
          if (error) {
            validationErrors[key as keyof T] = error;
          }
        }
      }
    }

    return validationErrors;
  };

  const values = reactive(initialValues);
  const errors = reactive(validate(initialValues));
  const touched = reactive({} as Partial<Record<keyof T, boolean | Record<string, boolean>>>);

  const setValues = (newValues: T) => {
    Object.assign(values, newValues); // Safely copy values into reactive object
    Object.assign(errors, validate(newValues)); // Validate updated values
  };

  const setErrors = (newErrors: T) => {
    Object.assign(errors, newErrors);
  };

  const reset = () => {
    Object.assign(values, initialValues);
    Object.assign(errors, validate(initialValues));
  };

  const setFieldValue = (field: string, value: unknown) => {
    updateNestedProperty(values as Record<string, unknown>, field, value);
  };

  const setFieldTouched = (field: string, touchedValue: boolean) => {
    updateNestedProperty(touched as Record<string, unknown>, field, touchedValue);
  }

  const handleSubmit = (e: Event) => {
    e.preventDefault();
    onSubmit(
      toRaw(values) as T,
      {
        setErrors,
        reset,
        setValues,
      } as FormikHelpers<T>
    );
  };

  const handleBlur = (e: FocusEvent) => {
    const fieldName = (e.target as HTMLInputElement).name as keyof T;
    setFieldTouched(fieldName.toString(), true);
  }
  const handleChange = (e: Event) => {
    const target = e.target as HTMLInputElement;
    const fieldName = target.name as keyof T;
    const value = target.type === 'checkbox' ? target.checked : target.value;

    setFieldValue(fieldName.toString(), value);
  }

  const isValid = computed(() => {
    return Object.keys(errors).length === 0;
  })

  const isDirty = computed(() => {
    return JSON.stringify(values) !== JSON.stringify(initialValues);
  })

  watch(
    () => values,
    () => {
      const validationErrors = validate(toRaw(values) as T);
      //@ts-expect-error Object.keys(errors) is not iterable
      Object.keys(errors).forEach((key) => delete errors[key]); // Clear existing errors
      Object.assign(errors, validationErrors); // Assign new errors
    },
    { deep: true }
  );


  const hasFieldError = (field: string) => {
    const errorValue = getNestedValue(errors, field);
    const touchedValue = getNestedValue(touched, field);
    return (errorValue !== undefined && !!touchedValue);
  }

  const getFieldError = (field: string) => {
    if (hasFieldError(field)) {
      return errors[field as keyof typeof errors];
    } else {
      return '';
    }
  }

  return {
    values,
    errors,
    touched,
    setValues,
    reset,
    setFieldValue,
    setFieldTouched,
    handleBlur,
    handleChange,
    handleSubmit,
    hasFieldError,
    getFieldError,
    isValid,
    isDirty
  };
};

export default useFormik;
