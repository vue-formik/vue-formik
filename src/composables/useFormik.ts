import { computed, reactive, toRaw, watch } from 'vue'

interface FormikHelpers<T> {
  setErrors: (errors: Partial<Record<keyof T, string | Record<string, string>>>) => void;
  reset: () => void;
  setValues: (values: T) => void;
}

interface ValidationRule<T> {
  (value: T): string | undefined;
}

interface IUseFormik<T> {
  initialValues: T;
  validationSchema: Record<keyof T, ValidationRule<T[keyof T]> | Record<string, ValidationRule<unknown>>>;
  onSubmit: (values: T, helpers: FormikHelpers<T>) => void;
}

// Utility function to update nested properties
const updateNestedProperty = (
  object: Record<string, unknown>,
  path: string,
  value: unknown
): void => {
  const keys = path.split('.');
  keys.reduce<Record<string, unknown>>((acc, key, index) => {
    if (index === keys.length - 1) {
      (acc as Record<string, unknown>)[key] = value;
    } else {
      if (!(acc as Record<string, unknown>)[key]) {
        (acc as Record<string, unknown>)[key] = {};
      }
      return (acc as Record<string, unknown>)[key] as Record<string, unknown>;
    }
    return acc;
  }, object);
};

// Validation function
const validate = <T>(
  values: T,
  schema: IUseFormik<T>['validationSchema']
): Partial<Record<keyof T, string>> => {
  const errors: Partial<Record<keyof T, string>> = {}; // Clear errors before starting validation

  for (const key in schema) {
    const value = values[key as keyof T];
    const rule = schema[key as keyof T] as ValidationRule<T[keyof T]>;

    if (typeof rule === 'function') {
      const error = rule(value);
      if (error) {
        errors[key as keyof T] = error;
      }
    }
  }

  return errors; // Only return the new validation results
};

const useFormik = <T extends object>({
  initialValues,
  validationSchema,
  onSubmit,
}: IUseFormik<T>) => {
  const values = reactive(initialValues);
  const errors = reactive(validate(initialValues, validationSchema));
  const touched = reactive({} as Partial<Record<keyof T, boolean | Record<string, boolean>>>);

  const setValues = (newValues: T) => {
    Object.assign(values, newValues); // Safely copy values into reactive object
    Object.assign(errors, validate(newValues, validationSchema)); // Validate updated values
  };

  const setErrors = (newErrors: T) => {
    Object.assign(errors, newErrors);
  };

  const reset = () => {
    Object.assign(values, initialValues);
    Object.assign(errors, validate(initialValues, validationSchema));
  };

  const setFieldValue = (field: string, value: unknown) => {
    updateNestedProperty(values as Record<string, unknown>, field, value);
  };

  const setFieldTouched = (field: string, touchedValue: boolean) => {
    updateNestedProperty(touched as Record<string, unknown>, field, touchedValue);
  }

  const handleSubmit = (e: Event) => {
    e.preventDefault();
    onSubmit(
      toRaw(values) as T,
      {
        setErrors,
        reset,
        setValues,
      } as FormikHelpers<T>
    );
  };

  const handleBlur = (e: FocusEvent) => {
    const fieldName = (e.target as HTMLInputElement).name as keyof T;
    setFieldTouched(fieldName.toString(), true);
  }
  const handleChange = (e: Event) => {
    const target = e.target as HTMLInputElement;
    const fieldName = target.name as keyof T;
    const value = target.type === 'checkbox' ? target.checked : target.value;

    setFieldValue(fieldName.toString(), value);
  }

  const isValid = computed(() => {
    return Object.keys(errors).length === 0;
  })

  const isDirty = computed(() => {
    return JSON.stringify(values) !== JSON.stringify(initialValues);
  })

  watch(
    () => values,
    () => {
      const validationErrors = validate(toRaw(values) as T, validationSchema);
      //@ts-expect-error Object.keys(errors) is not iterable
      Object.keys(errors).forEach((key) => delete errors[key]); // Clear existing errors
      Object.assign(errors, validationErrors); // Assign new errors
    },
    { deep: true }
  );

  return {
    values,
    errors,
    touched,
    setValues,
    reset,
    setFieldValue,
    setFieldTouched,
    handleBlur,
    handleChange,
    handleSubmit,
    isValid,
    isDirty
  };
};

export default useFormik;
