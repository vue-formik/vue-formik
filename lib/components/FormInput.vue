<template>
  <div
    :class="{
      'vf-field': true,
      [typeClass]: true,
      'vf-field--error': hasError,
    }"
  >
    <label v-if="label" :for="name" :id="name + '-label'">
      {{ label }}
    </label>

    <div class="vf-input">
      <slot name="prepend" />

      <input
        :id="name"
        :name="name"
        :type="type || 'text'"
        :placeholder="placeholder"
        :value="inputValue"
        :readonly="readonly"
        :required="required"
        :disabled="disabled"
        @input="handleInput"
        @blur="formik.handleFieldBlur"
        :class="{
          'vf-input--error': hasError,
          'vf-input--readonly': readonly,
          'vf-input--disabled': disabled,
        }"
        v-bind="inputProps"
        :aria-labelledby="label ? name + '-label' : undefined"
        :aria-describedby="hasError ? name + '-error' : undefined"
        :aria-invalid="hasError ? 'true' : 'false'"
        :aria-required="required ? 'true' : undefined"
        :aria-readonly="readonly ? 'true' : undefined"
        :aria-disabled="disabled ? 'true' : undefined"
      />

      <slot name="append" />
    </div>

    <template v-if="hasError">
      <p
        v-if="typeof getError === 'string'"
        class="vf-error" :id="name + '-error'"
      >
        {{ getError }}
      </p>
      <template v-else-if="Array.isArray(getError)">
        <ul class="vf-error" :id="name + '-error'">
          <li v-for="(error, index) in getError" :key="index">
            {{ error }}
          </li>
        </ul>
      </template>
    </template>

    <slot />
  </div>
</template>

<script lang="ts" setup>
/* eslint-disable @typescript-eslint/no-explicit-any */
import useFormik from "@/composables/useFormik";
import { computed } from "vue";

const props = defineProps<{
  formik: ReturnType<typeof useFormik<any>>;
  name: string;
  label?: string;
  type?: string;
  placeholder?: string;
  readonly?: boolean;
  disabled?: boolean;
  required?: boolean;
  inputProps?: Record<keyof HTMLInputElement, any>;
}>();

const inputValue = computed(() => props.formik.getFieldValue(props.name) as string);

const handleInput = (e: Event) => {
  const value = (e.target as HTMLInputElement).value;
  props.formik.setFieldValue(props.name, value);
};

const typeClass = computed(() => `vf-${props.type || "text"}-field`);

const hasError = computed(() => props.formik.hasFieldError(props.name));
const getError = computed(() => props.formik.getFieldError(props.name));
</script>
