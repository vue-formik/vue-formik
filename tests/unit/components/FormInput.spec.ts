import { describe, test, expect, beforeEach } from "vitest";

import { mount, shallowMount, VueWrapper } from "@vue/test-utils";
import FormInput from "@/components/FormInput.vue";
import useFormik from "@/composables/useFormik";
import { nextTick } from "vue";

describe("FormInput", async () => {
  const formik = useFormik({
    initialValues: {
      fullName: "",
      addresses: [""],
      contact: {
        code: "",
        number: "",
      },
    },
    validationSchema: {
      fullName: (value: string) => {
        if (!value) {
          return "Full Name is required";
        }
        if (value.length < 5) {
          return "Full Name is too short";
        }
        if (/[^a-zA-Z ]/.test(value)) {
          return "Full Name is invalid";
        }
      },
      addresses: (values: string[]) => {
        if (!values.length) {
          return "Addresses is required";
        }
        const errors: string[][] = [];
        values.forEach((value) => {
          const e: string[] = [];
          if (!value) {
            e.push("Address is required");
          }
          if (value.length < 5) {
            e.push("Address is too short");
          }
          if (/[^a-zA-Z0-9 ]/.test(value)) {
            e.push("Address is invalid");
          }

          errors.push(e);
        });
        return errors;
      },
      contact: (value: { code: string; number: string }) => {
        const err = {
          code: "",
          number: "",
        };
        if (!value.code) {
          err.code = "Code is required";
        }
        if (value.code.length < 3) {
          err.code = "Code is too short";
        }
        if (!value.number) {
          err.number = "Number is required";
        }
        if (value.number.length < 5) {
          err.number = "Number is too short";
        }
        return err;
      },
    },
  });

  let wrapper: VueWrapper<unknown>;

  describe("formik prop usage", () => {
    beforeEach(() => {
      formik.reset();
      wrapper = mount(FormInput, {
        props: {
          name: "fullName",
          formik,
        },
      });
    });

    test("renders input", () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    test("handleChange", () => {
      wrapper.find("input").setValue("John Doe");
      expect(formik.values.fullName).toBe("John Doe");
      expect(wrapper.find("input").element.value).toBe("John Doe");
    });

    test("renders label", async () => {
      await wrapper.setProps({ label: "Full Name" });
      await nextTick();
      expect(wrapper.html()).toMatchSnapshot();
    });

    test("shows error message on blur", async () => {
      await wrapper.find("input").setValue("");
      await wrapper.find("input").trigger("blur");
      await nextTick();
      expect(wrapper.html()).toMatchSnapshot();
    });

    describe("Field with array index", () => {
      beforeEach(() => {
        wrapper = mount(FormInput, {
          props: {
            name: "addresses[0]",
            formik: formik,
          },
        });
      });

      test("renders input", () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      test("various errors", async () => {
        await wrapper.find("input").trigger("blur");
        await nextTick();
        expect(wrapper.html()).toMatchSnapshot();
        await wrapper.find("input").setValue("$%");
        expect(wrapper.html()).toMatchSnapshot();
      });

      test("handleChange", () => {
        wrapper.find("input").setValue("123 Main St");
        expect(formik.values.addresses[0]).toBe("123 Main St");
        expect(wrapper.find("input").element.value).toBe("123 Main St");
      });
    });

    describe("Field with object type", () => {
      beforeEach(() => {
        wrapper = mount(FormInput, {
          props: {
            name: "contact.code",
            formik: formik,
          },
        });
      });

      test("renders input", () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      test("various errors", async () => {
        await wrapper.find("input").trigger("blur");
        await nextTick();
        expect(wrapper.html()).toMatchSnapshot();
        await wrapper.find("input").setValue("123");
        expect(wrapper.html()).toMatchSnapshot();
      });

      test("handleChange", () => {
        wrapper.find("input").setValue("123");
        expect(formik.values.contact.code).toBe("123");
        expect(wrapper.find("input").element.value).toBe("123");
      });
    });

    describe("type class", () => {
      test("default", () => {
        expect(wrapper.classes()).toContain("vf-text-field");
      });
      test("password", async () => {
        await wrapper.setProps({ type: "password" });
        expect(wrapper.classes()).toContain("vf-password-field");
        expect(wrapper.classes()).not.toContain("vf-text-field");
      });
    });

    test("required prop", async () => {
      await wrapper.setProps({ required: true });
      await nextTick();
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe("provide formik and inject usage", () => {
    beforeEach(() => {
      formik.reset();
      wrapper = shallowMount(FormInput, {
        props: {
          name: "fullName",
        },
        global: {
          provide: {
            formik: formik,
          },
        },
      });
    });

    test("handleChange", async () => {
      await wrapper.find("input").setValue("John Doe");
      await nextTick();
      expect(formik.values.fullName).toBe("John Doe");
      expect(wrapper.find("input").element.value).toBe("John Doe");
    });
  });
});
