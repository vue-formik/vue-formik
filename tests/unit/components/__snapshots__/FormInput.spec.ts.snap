// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormInput > Field with array index > renders input 1`] = `
"<div class="vf-field vf-text-field">
  <!--v-if-->
  <div class="vf-input"><input id="addresses[0]" name="addresses[0]" type="text" class="" aria-invalid="false" value=""></div>
  <!--v-if-->
</div>"
`;

exports[`FormInput > Field with array index > various errors 1`] = `
"<div class="vf-field vf-text-field vf-field--error">
  <!--v-if-->
  <div class="vf-input"><input id="addresses[0]" name="addresses[0]" type="text" class="vf-input--error" aria-invalid="true" value="" aria-describedby="addresses[0]-error"></div>
  <ul class="vf-error" id="addresses[0]-error">
    <li>Address is required</li>
    <li>Address is too short</li>
  </ul>
</div>"
`;

exports[`FormInput > Field with array index > various errors 2`] = `
"<div class="vf-field vf-text-field vf-field--error">
  <!--v-if-->
  <div class="vf-input"><input id="addresses[0]" name="addresses[0]" type="text" class="vf-input--error" aria-invalid="true" value="$%" aria-describedby="addresses[0]-error"></div>
  <ul class="vf-error" id="addresses[0]-error">
    <li>Address is too short</li>
    <li>Address is invalid</li>
  </ul>
</div>"
`;

exports[`FormInput > Field with object type > renders input 1`] = `
"<div class="vf-field vf-text-field">
  <!--v-if-->
  <div class="vf-input"><input id="contact.code" name="contact.code" type="text" class="" aria-invalid="false" value=""></div>
  <!--v-if-->
</div>"
`;

exports[`FormInput > Field with object type > various errors 1`] = `
"<div class="vf-field vf-text-field vf-field--error">
  <!--v-if-->
  <div class="vf-input"><input id="contact.code" name="contact.code" type="text" class="vf-input--error" aria-invalid="true" value="" aria-describedby="contact.code-error"></div>
  <p class="vf-error" id="contact.code-error">Code is too short</p>
</div>"
`;

exports[`FormInput > Field with object type > various errors 2`] = `
"<div class="vf-field vf-text-field">
  <!--v-if-->
  <div class="vf-input"><input id="contact.code" name="contact.code" type="text" class="" aria-invalid="false" value="123"></div>
  <!--v-if-->
</div>"
`;

exports[`FormInput > formik prop usage > Field with array index > renders input 1`] = `
"<div class="vf-field vf-text-field">
  <!--v-if-->
  <div class="vf-input"><input id="addresses[0]" name="addresses[0]" type="text" class="" aria-invalid="false" value=""></div>
  <!--v-if-->
</div>"
`;

exports[`FormInput > formik prop usage > Field with array index > various errors 1`] = `
"<div class="vf-field vf-text-field vf-field--error">
  <!--v-if-->
  <div class="vf-input"><input id="addresses[0]" name="addresses[0]" type="text" class="vf-input--error" aria-invalid="true" value="" aria-describedby="addresses[0]-error"></div>
  <ul class="vf-error" id="addresses[0]-error">
    <li>Address is required</li>
    <li>Address is too short</li>
  </ul>
</div>"
`;

exports[`FormInput > formik prop usage > Field with array index > various errors 2`] = `
"<div class="vf-field vf-text-field vf-field--error">
  <!--v-if-->
  <div class="vf-input"><input id="addresses[0]" name="addresses[0]" type="text" class="vf-input--error" aria-invalid="true" value="$%" aria-describedby="addresses[0]-error"></div>
  <ul class="vf-error" id="addresses[0]-error">
    <li>Address is too short</li>
    <li>Address is invalid</li>
  </ul>
</div>"
`;

exports[`FormInput > formik prop usage > Field with object type > renders input 1`] = `
"<div class="vf-field vf-text-field">
  <!--v-if-->
  <div class="vf-input"><input id="contact.code" name="contact.code" type="text" class="" aria-invalid="false" value=""></div>
  <!--v-if-->
</div>"
`;

exports[`FormInput > formik prop usage > Field with object type > various errors 1`] = `
"<div class="vf-field vf-text-field vf-field--error">
  <!--v-if-->
  <div class="vf-input"><input id="contact.code" name="contact.code" type="text" class="vf-input--error" aria-invalid="true" value="" aria-describedby="contact.code-error"></div>
  <p class="vf-error" id="contact.code-error">Code is too short</p>
</div>"
`;

exports[`FormInput > formik prop usage > Field with object type > various errors 2`] = `
"<div class="vf-field vf-text-field">
  <!--v-if-->
  <div class="vf-input"><input id="contact.code" name="contact.code" type="text" class="" aria-invalid="false" value="123"></div>
  <!--v-if-->
</div>"
`;

exports[`FormInput > formik prop usage > renders input 1`] = `
"<div class="vf-field vf-text-field">
  <!--v-if-->
  <div class="vf-input"><input id="fullName" name="fullName" type="text" class="" aria-invalid="false" value=""></div>
  <!--v-if-->
</div>"
`;

exports[`FormInput > formik prop usage > renders label 1`] = `
"<div class="vf-field vf-text-field"><label for="fullName" id="fullName-label">Full Name</label>
  <div class="vf-input"><input id="fullName" name="fullName" type="text" class="" aria-invalid="false" value="" aria-labelledby="fullName-label"></div>
  <!--v-if-->
</div>"
`;

exports[`FormInput > formik prop usage > required prop 1`] = `
"<div class="vf-field vf-text-field">
  <!--v-if-->
  <div class="vf-input"><input id="fullName" name="fullName" type="text" class="" aria-invalid="false" value="" required="" aria-required="true"></div>
  <!--v-if-->
</div>"
`;

exports[`FormInput > formik prop usage > shows error message on blur 1`] = `
"<div class="vf-field vf-text-field vf-field--error">
  <!--v-if-->
  <div class="vf-input"><input id="fullName" name="fullName" type="text" class="vf-input--error" aria-invalid="true" value="" aria-describedby="fullName-error"></div>
  <p class="vf-error" id="fullName-error">Full Name is required</p>
</div>"
`;

exports[`FormInput > renders input 1`] = `
"<div class="vf-field vf-text-field">
  <!--v-if-->
  <div class="vf-input"><input id="fullName" name="fullName" type="text" class="" aria-invalid="false" value=""></div>
  <!--v-if-->
</div>"
`;

exports[`FormInput > renders label 1`] = `
"<div class="vf-field vf-text-field"><label for="fullName" id="fullName-label">Full Name</label>
  <div class="vf-input"><input id="fullName" name="fullName" type="text" class="" aria-invalid="false" value="" aria-labelledby="fullName-label"></div>
  <!--v-if-->
</div>"
`;

exports[`FormInput > required prop 1`] = `
"<div class="vf-field vf-text-field">
  <!--v-if-->
  <div class="vf-input"><input id="fullName" name="fullName" type="text" class="" aria-invalid="false" value="" required="" aria-required="true"></div>
  <!--v-if-->
</div>"
`;

exports[`FormInput > shows error message on blur 1`] = `
"<div class="vf-field vf-text-field vf-field--error">
  <!--v-if-->
  <div class="vf-input"><input id="fullName" name="fullName" type="text" class="vf-input--error" aria-invalid="true" value="" aria-describedby="fullName-error"></div>
  <p class="vf-error" id="fullName-error">Full Name is required</p>
</div>"
`;
